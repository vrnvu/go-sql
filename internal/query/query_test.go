package query

import (
	"encoding/csv"
	"os"
	"testing"
	"time"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
)

func TestQuerySnapshot(t *testing.T) {
	if testing.Short() {
		t.Skip("slow: query snapshot")
	}

	t.Parallel()
	startTime := time.Date(2025, 1, 1, 0, 0, 0, 0, time.UTC)
	endTime := time.Date(2025, 1, 1, 0, 0, 1, 0, time.UTC)

	query := Query{
		Hostname:  "host1",
		StartTime: startTime,
		EndTime:   endTime,
	}

	snaps.MatchSnapshot(t, query.Build())
}

func AssertHeaders(t *testing.T, fields []string) {
	assert.Equal(t, 3, len(fields))
	assert.Equal(t, "hostname", fields[0])
	assert.Equal(t, "start_time", fields[1])
	assert.Equal(t, "end_time", fields[2])
}

func TestFieldsSnapshot(t *testing.T) {
	t.Parallel()
	inputFile, err := os.Open("../../resources/query_params.csv")
	assert.NoError(t, err)
	defer inputFile.Close()

	reader := csv.NewReader(inputFile)
	fields, err := reader.Read()
	assert.NoError(t, err)
	AssertHeaders(t, fields)
}

func TestQueryInvalidHeaderSnapshot(t *testing.T) {
	t.Parallel()
	inputFile, err := os.Open("../../resources/invalid_headers.csv")
	assert.NoError(t, err)
	defer inputFile.Close()

	reader := csv.NewReader(inputFile)
	queryReader, err := NewQueryReader(reader)
	assert.Error(t, err)
	assert.Nil(t, queryReader)
	snaps.MatchSnapshot(t, err.Error())
}

func TestQueryInvalidRowSnapshot(t *testing.T) {
	t.Parallel()
	inputFile, err := os.Open("../../resources/invalid_row.csv")
	assert.NoError(t, err)
	defer inputFile.Close()

	// line 1: valid headers
	reader := csv.NewReader(inputFile)
	queryReader, err := NewQueryReader(reader)
	assert.NoError(t, err)
	assert.NotNil(t, queryReader)

	// line 2: valid
	query, hasMore, err := queryReader.Next()
	assert.NoError(t, err)
	assert.NotNil(t, query)
	assert.True(t, hasMore)
	snaps.MatchSnapshot(t, query.Build())

	// line 3: invalid value
	query, hasMore, err = queryReader.Next()
	assert.Error(t, err)
	assert.Empty(t, query)
	assert.True(t, hasMore)
	snaps.MatchSnapshot(t, err.Error())

	// line 4: extra field
	query, hasMore, err = queryReader.Next()
	assert.Error(t, err)
	assert.Empty(t, query)
	assert.False(t, hasMore)
	snaps.MatchSnapshot(t, err.Error())
}
