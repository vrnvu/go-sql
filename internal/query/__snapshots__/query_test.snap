
[TestQuerySnapshot - 1]
SELECT * FROM cpu_usage WHERE host = 'host1' AND ts BETWEEN '2025-01-01 00:00:00' AND '2025-01-01 00:00:01'
---

[TestQueryInvalidHeaderSnapshot - 1]
expected 3 fields, got 2
---

[TestQueryInvalidRowSnapshot - 1]
SELECT * FROM cpu_usage WHERE host = 'host_000001' AND ts BETWEEN '2017-01-01 08:59:22' AND '2017-01-01 09:59:22'
---

[TestQueryInvalidRowSnapshot - 2]
invalid start_time: INVALID TIME HERE err: parsing time "INVALID TIME HERE" as "2006-01-02 15:04:05": cannot parse "INVALID TIME HERE" as "2006" on line 3
---

[TestQueryInvalidRowSnapshot - 3]
error reading CSV record: record on line 4: wrong number of fields on line 4
---
